File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/bugs/bug_status_controller.py
from django.contrib import messages
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from ....Models.bugs.bug_status import BugStatus
from .._traits.controller import Controller
from .._helpers.http import get_redirect_url
from django.core.exceptions import PermissionDenied
from ....configs.messages_templates import get_exception_class_message

class BugStatusController(Controller):

  def index(self, request: HttpRequest) -> HttpResponse:
    try:
      bug_status = BugStatus.objects.filter(created_by=request.user.creator_id()).order_by('order')
      return render(request, 'bugstatus/index.html', {'bugStatus': bug_status})
    except Exception as e:
      print(f"[BugStatusController.index] Failed to load bug statuses: {e.__class__.__name__}: {e}")
      messages.error(request, "Error loading bug statuses.")
      return redirect(get_redirect_url(request))

  def create(self, request: HttpRequest) -> HttpResponse:
    try:
      return render(request, 'bugstatus/create.html', {})
    except Exception as e:
      print(f"[BugStatusController.create] Failed to load create view: {e.__class__.__name__}: {e}")
      messages.error(request, "Error loading create view.")
      return redirect(get_redirect_url(request))

  def store(self, request: HttpRequest) -> HttpResponse:
    try:
      if request.method == 'POST':
        title = request.POST.get('title')
        if not title or len(title) > 20:
          messages.error(request, "Title is required and must be at most 20 characters.")
          return redirect('bugstatus_index')
        all_status = BugStatus.objects.filter(created_by=request.user.creator_id()).order_by('-id').first()
        order_value = all_status.order + 1 if all_status else 0
        status = BugStatus(title=title, created_by=request.user.creator_id(), order=order_value)
        status.save()
        messages.success(request, "Bug status successfully created.")
        return redirect('bugstatus_index')
      else:
        messages.error(request, "Invalid request method.")
        return redirect('bugstatus_index')
    except Exception as e:
      print(f"[BugStatusController.store] Failed to store bug status: {e.__class__.__name__}: {e}")
      messages.error(request, "Error storing bug status.")
      return redirect(get_redirect_url(request))

  def edit(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      bug_status = get_object_or_404(BugStatus, pk=id)
      return (render(request, 'bugstatus/edit.html', {'bugStatus': bug_status})
              if bug_status.created_by == request.user.creator_id()
              else JsonResponse({'error': "Permission denied."}, status=401))
    except Exception as e:
      print(f"[BugStatusController.edit] Failed to load edit view: {e.__class__.__name__}: {e}")
      messages.error(request, "Error loading edit view.")
      return redirect(get_redirect_url(request))

  def update(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      bug_status = get_object_or_404(BugStatus, pk=id)
      if bug_status.created_by == request.user.creator_id():
        if request.method == 'POST':
          title = request.POST.get('title')
          if not title or len(title) > 20:
            messages.error(request, "Title is required and must be at most 20 characters.")
            return redirect('bugstatus_index')
          bug_status.title = title
          bug_status.save()
          messages.success(request, "Bug status successfully updated.")
          return redirect('bugstatus_index')
        else:
          messages.error(request, "Invalid request method.")
          return redirect('bugstatus_index')
      else:
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
    except Exception as e:
      print(f"[BugStatusController.update] Failed to update bug status: {e.__class__.__name__}: {e}")
      messages.error(request, "Error updating bug status.")
      return redirect(get_redirect_url(request))

  def destroy(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      bug_status = get_object_or_404(BugStatus, pk=id)
      if bug_status.created_by == request.user.creator_id():
        bug_status.delete()
        messages.success(request, "Bug status successfully deleted.")
        return redirect('bugstatus_index')
      else:
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
    except Exception as e:
      print(f"[BugStatusController.destroy] Failed to delete bug status: {e.__class__.__name__}: {e}")
      messages.error(request, "Error deleting bug status.")
      return redirect(get_redirect_url(request))

  def order(self, request: HttpRequest) -> JsonResponse:
    try:
      if request.method == 'POST':
        order_list = request.POST.getlist('order')
        for key, item in enumerate(order_list):
          status = BugStatus.objects.filter(id=item).first()
          if status:
            status.order = key
            status.save()
        return JsonResponse({'success': True})
      else:
        return JsonResponse({'error': "Invalid request method."}, status=400)
    except Exception as e:
      print(f"[BugStatusController.order] Failed to update order: {e.__class__.__name__}: {e}")
      return JsonResponse({'error': "Error updating order."}, status=500)

File: _inc/laravel/app/Http/Controllers/Bugs/BugStatusController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    ActivitiesConstants,
    DatabaseConstants,
    MiddlewaresConstants,
    PermissionsConstants,
    UsersConstants,
    ViewsConstants
};
use App\{
    Models\BugStatus,
    Traits\ChecksLogin
};
use App\Services\BugReportService;
use App\Traits\ChecksPermissions;
use Illuminate\{
    Auth\Access\AuthorizationException,
    Support\Facades\Log,
    Validation\ValidationException,
    View\View
};
use Illuminate\Database\{
    Eloquent\ModelNotFoundException,
    QueryException
};
use Illuminate\Http\{
    JsonResponse,
    RedirectResponse,
    Request
};
use Illuminate\Support\Facades\View as ViewFacade;
use Symfony\Component\HttpFoundation\Response;
use Throwable;

final class BugStatusController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    private BugReportService $bugReportService;
    public function __construct(BugReportService $bugReportService)
    {
        $this->middleware(MiddlewaresConstants::AUTH);
        $this->bugReportService = $bugReportService;
    }

    public function index(Request $request): View|RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, PermissionsConstants::MNG_BUG_STT, ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $creatorId = $request->user()?->creatorId();
                $bugStatuses = BugStatus::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)
                    ->orderBy(ActivitiesConstants::COL_OD)
                    ->get();

                $view = ViewsConstants::BUG_STT . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \Exception('view'), $action, route(ViewsConstants::BUG_STT . '.index'));

                return ViewFacade::make($view, ['bug_statuses' => $bugStatuses]);
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@index authorization failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_FORBIDDEN, 'Forbidden');
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@index failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Unable to load bug statuses');
            }
        }, [UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
    }

    public function create(Request $request): View|RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, 'create bug status', ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $view = ViewsConstants::BUG_STT . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \Exception('view'), $action, route(ViewsConstants::BUG_STT . '.index'));

                return ViewFacade::make($view);
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@create authorization failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_FORBIDDEN, 'Forbidden');
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@create failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Unable to show creation form');
            }
        }, [UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
    }

    public function store(Request $request): RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, 'create bug status', ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $data = $request->validate([
                    ActivitiesConstants::COL_TT => 'required|string|max:20',
                ]);

                $creatorId = $request->user()->creatorId();
                $maxOrder = BugStatus::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)->max(ActivitiesConstants::COL_OD);

                BugStatus::create([
                    ActivitiesConstants::COL_TT      => $data[ActivitiesConstants::COL_TT],
                    ActivitiesConstants::COL_OD      => ($maxOrder ?? -1) + 1,
                    DatabaseConstants::COL_TABLE_CREATOR => $creatorId,
                ]);

                return redirect()->route(ViewsConstants::BUG_STT . '.index')->with('success', __('Bug status successfully created.'));
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@store authorization failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_FORBIDDEN, 'Forbidden');
            } catch (ValidationException $e) {
                Log::warning('BugStatusController@store validation failed', ['errors' => $e->errors()]);
                return $this->errorResponse($request, Response::HTTP_UNPROCESSABLE_ENTITY, $e->validator->errors()->first(), $e->errors());
            } catch (QueryException $e) {
                Log::error('BugStatusController@store database error', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Database error');
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@store failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Unable to create bug status');
            }
        }, [UsersConstants::COL_USER_ID => $request->user()?->id ?? null, 'input' => $request->only(ActivitiesConstants::COL_TT)]);
    }

    public function edit(Request $request, int|string $id): View|RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $id, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, 'edit bug status', ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $bugStatus = BugStatus::findOrFail($id);
                if (!self::isOwner($bugStatus)) throw new AuthorizationException('You do not own this resource');

                $view = ViewsConstants::BUG_STT . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \Exception('view'), $action, route(ViewsConstants::BUG_STT . '.index'));

                return ViewFacade::make($view, ['bug_status' => $bugStatus]);
            } catch (ModelNotFoundException $e) {
                Log::warning('BugStatusController@edit not found', ['id' => $id]);
                return $this->errorResponse($request, Response::HTTP_NOT_FOUND, 'Bug status not found');
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@edit authorization failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_FORBIDDEN, 'Forbidden');
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@edit failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Unable to load edit form');
            }
        }, ['id' => $id]);
    }

    public function update(Request $request, int|string $id): RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $id, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, 'edit bug status', ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $bugStatus = BugStatus::findOrFail($id);
                if (!self::isOwner($bugStatus)) throw new AuthorizationException('You do not own this resource');

                $data = $request->validate([
                    ActivitiesConstants::COL_TT => 'required|string|max:20',
                ]);

                $bugStatus->update([ActivitiesConstants::COL_TT => $data[ActivitiesConstants::COL_TT]]);

                return redirect()->route(ViewsConstants::BUG_STT . '.index')->with('success', __('Bug status successfully updated.'));
            } catch (ModelNotFoundException $e) {
                Log::warning('BugStatusController@update not found', ['id' => $id]);
                return $this->errorResponse($request, Response::HTTP_NOT_FOUND, 'Bug status not found');
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@update authorization failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_FORBIDDEN, 'Forbidden');
            } catch (ValidationException $e) {
                Log::warning('BugStatusController@update validation failed', ['errors' => $e->errors()]);
                return $this->errorResponse($request, Response::HTTP_UNPROCESSABLE_ENTITY, $e->validator->errors()->first(), $e->errors());
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@update failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Unable to update bug status');
            }
        }, ['id' => $id, 'input' => $request->only(ActivitiesConstants::COL_TT)]);
    }

    public function destroy(Request $request, int|string $id): RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $id, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, 'delete bug status', ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $bugStatus = BugStatus::findOrFail($id);
                if (!self::isOwner($bugStatus)) throw new AuthorizationException('You do not own this resource');

                $bugStatus->delete();

                return redirect()->route(ViewsConstants::BUG_STT . '.index')->with('success', __('Bug status successfully deleted.'));
            } catch (ModelNotFoundException $e) {
                Log::warning('BugStatusController@destroy not found', ['id' => $id]);
                return $this->errorResponse($request, Response::HTTP_NOT_FOUND, 'Bug status not found');
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@destroy authorization failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_FORBIDDEN, 'Forbidden');
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@destroy failed', ['exception' => $e]);
                return $this->errorResponse($request, Response::HTTP_INTERNAL_SERVER_ERROR, 'Unable to delete bug status');
            }
        }, ['id' => $id]);
    }

    public function order(Request $request): Response|RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $cls, $meth, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($denial = self::guard($request, 'edit bug status', ViewsConstants::BUG_STT . '.index')) !== true) return $denial;

            try {
                $positions = $request->input(ActivitiesConstants::COL_OD, []);
                foreach ($positions as $index => $id) {
                    BugStatus::where('id', $id)
                        ->where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()->creatorId())
                        ->update([ActivitiesConstants::COL_OD => $index]);
                }
                return response()->noContent(Response::HTTP_OK);
            } catch (AuthorizationException $e) {
                Log::warning('BugStatusController@order authorization failed', ['exception' => $e]);
                return response()->json(['error' => 'Forbidden'], Response::HTTP_FORBIDDEN);
            } catch (\Throwable $e) {
                Log::critical('BugStatusController@order failed', ['exception' => $e]);
                return response()->json(['error' => 'Unable to reorder statuses'], Response::HTTP_INTERNAL_SERVER_ERROR);
            }
        }, [UsersConstants::COL_USER_ID => $request->user()?->id ?? null, ActivitiesConstants::COL_OD => $request->input(ActivitiesConstants::COL_OD, [])]);
    }

    public function show(): RedirectResponse
    {
        return redirect()->route(ViewsConstants::BUG_STT . '.index');
    }

    private static function deny(Request $request, string $permission): RedirectResponse|JsonResponse|null
    {
        if ($request->user()->can($permission))
            return null;
        return defaultPermissionDenial(
            $request,
            new AuthorizationException($permission),
            __CLASS__ . '::' . debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1]['function']
        );
    }

    private static function isOwner(BugStatus $status): bool
    {
        $userOrRedirect = self::_checkLogin();
        if ($userOrRedirect instanceof RedirectResponse)
            return false;
        /** @var \App\Models\User $user */
        $user = $userOrRedirect;
        return $status[DatabaseConstants::COL_TABLE_CREATOR] === $user?->creatorId();
    }

    /**
     * Build either a JSON error or a RedirectResponse + flash
     *
     * @param  Request      $request
     * @param  int          $statusCode
     * @param  string       $message
     * @param  array|null   $errors
     * @return JsonResponse|RedirectResponse
     */
    private function errorResponse(
        Request $request,
        int     $statusCode,
        string  $message,
        ?array  $errors = null
    ): JsonResponse|RedirectResponse {
        if ($request->expectsJson()) {
            $payload = ['message' => $message];
            if ($errors)
                $payload['errors'] = $errors;
            return response()->json($payload, $statusCode);
        }
        return redirect()->back()
            ->withErrors($errors ?? ['error' => $message])
            ->withInput();
    }
}
